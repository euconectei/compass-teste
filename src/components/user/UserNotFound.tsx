type UserNotFoundType = {
  username: string;
};

const UserNotFound = ({ username }: UserNotFoundType) => {
  return (
    <div className='alert alert-info'>
      O usuário {username} não foi encontrado.
    </div>
  );
};

export default UserNotFound;
