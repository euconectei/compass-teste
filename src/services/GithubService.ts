import { RepoType } from '../types/RepoType';
import { StarredType } from '../types/StarredType';
import { UserType } from '../types/UserType';
import instance from './instance';

const GithubService = {
  getRepos: async (user: string): Promise<RepoType[]> => {
    try {
      const res = await instance.get(`/users/${user}/repos`);
      return res.data;
    } catch (error) {
      return [];
    }
  },

  getStarred: async (user: string): Promise<StarredType[]> => {
    try {
      const res = await instance.get(`/users/${user}/starred`);
      return res.data;
    } catch (error) {
      return [];
    }
  },

  getUser: async (user: string): Promise<UserType> => {
    try {
      const res = await instance.get(`/users/${user}`);
      return res.data;
    } catch (error) {
      return {} as UserType;
    }
  },
};

export default GithubService;
