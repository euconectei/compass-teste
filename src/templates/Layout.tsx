import { Outlet } from 'react-router-dom';

import { useAuth } from '../contexts/AuthContext';
import HomeSignedOut from '../pages/HomeSignedOut';
import { useLoading } from '../contexts/LoadingContext';

const Layout = () => {
  const { isAuthenticated, user } = useAuth();
  const { isLoading, message } = useLoading();

  return (
    <>
      {isLoading && (
        <div className='loading_bg d-flex justify-content-center align-items-center'>
          <div className='loading text-center'>
            <div className='spinner-border' role='status'></div>
            <div>{message}</div>
          </div>
        </div>
      )}

      <div className='d-flex justify-content-center align-items-center flex-grow-1'>
        <div className='container'>
          <div className='row'>
            <div className='col-md-6 p-3'>
              <h1 className='layout-title'>Pesquise um Usu√°rio</h1>
            </div>
            <div className='col-md-6 p-3'>
              <main>{!isAuthenticated ? <Outlet /> : <HomeSignedOut />}</main>
            </div>
          </div>

          <div>{user}</div>
        </div>
      </div>
    </>
  );
};

export default Layout;
