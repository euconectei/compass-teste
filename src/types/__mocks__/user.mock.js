// 20220312225722
// https://api.github.com/users/euconectei
export const userMock = {
  login: 'euconectei',
  id: 1842710,
  node_id: 'MDQ6VXNlcjE4NDI3MTA=',
  avatar_url: 'https://avatars.githubusercontent.com/u/1842710?v=4',
  gravatar_id: '',
  url: 'https://api.github.com/users/euconectei',
  html_url: 'https://github.com/euconectei',
  followers_url: 'https://api.github.com/users/euconectei/followers',
  following_url:
    'https://api.github.com/users/euconectei/following{/other_user}',
  gists_url: 'https://api.github.com/users/euconectei/gists{/gist_id}',
  starred_url: 'https://api.github.com/users/euconectei/starred{/owner}{/repo}',
  subscriptions_url: 'https://api.github.com/users/euconectei/subscriptions',
  organizations_url: 'https://api.github.com/users/euconectei/orgs',
  repos_url: 'https://api.github.com/users/euconectei/repos',
  events_url: 'https://api.github.com/users/euconectei/events{/privacy}',
  received_events_url:
    'https://api.github.com/users/euconectei/received_events',
  type: 'User',
  site_admin: false,
  name: 'Rafael Fran√ßa',
  company: null,
  blog: '',
  location: 'Rio de Janeiro, RJ - Brasil',
  email: null,
  hireable: true,
  bio: null,
  twitter_username: null,
  public_repos: 40,
  public_gists: 4,
  followers: 35,
  following: 7,
  created_at: '2012-06-12T14:02:34Z',
  updated_at: '2022-03-03T12:45:39Z',
};
