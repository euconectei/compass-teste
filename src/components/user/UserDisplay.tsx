import { UserType } from '../../types/UserType';
import UserNotFound from './UserNotFound';

type UserDisplayComponentType = {
  user: UserType;
  username: string;
};

const UserDisplay = ({ user, username }: UserDisplayComponentType) => {
  if (Object.keys(user).length === 0) {
    return <UserNotFound username={username} />;
  }

  return (
    <>
      <div className='m-3'>
        <div className='card user-card'>
          <img
            src={user.avatar_url}
            alt={`Foto de ${user.login}`}
            className='card-img-top'
          />
          <div className='card-body'>
            <h5 className='card-title'>{user.login}</h5>
            <ul className='list-group list-group-flush'>
              <li className='list-group-item'>
                <span className='font-weight-bold'>Nome:</span> {user.name}
              </li>
              <li className='list-group-item'>
                <span className='font-weight-bold'>E-mail:</span> {user.email}
              </li>
              <li className='list-group-item'>
                <span className='font-weight-bold'>Reposit√≥rios:</span>{' '}
                {user.public_repos}
              </li>
              <li className='list-group-item'>
                <span className='font-weight-bold'>Seguidores:</span>{' '}
                {user.followers}
              </li>
              <li className='list-group-item'>
                <span className='font-weight-bold'>Seguindo:</span>{' '}
                {user.following}
              </li>
            </ul>
            <hr className='divider' />
            <div className='d-flex justify-content-between'>
              <a href={`/${user.login}/repos`} className='btn btn-primary'>
                Repos
              </a>
              <a href={`/${user.login}/starred`} className='btn btn-primary'>
                Starred
              </a>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};

export default UserDisplay;
