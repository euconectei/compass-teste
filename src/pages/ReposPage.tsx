import React, { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import GithubService from '../services/GithubService';
import { RepoType } from '../types/RepoType';
import { useLoading } from '../contexts/LoadingContext';
import RepoList from '../components/repo/RepoList';

const ReposPage = () => {
  const username = useParams().username;
  const [repos, setRepos] = useState<RepoType[]>([]);
  const navigate = useNavigate();
  const { setLoading } = useLoading();

  const handleNavigateUser = (ev: React.MouseEvent<HTMLAnchorElement>) => {
    ev.preventDefault();
    navigate(`/${username}`);
  };

  const getRepos = async () => {
    setLoading(true, 'Carregando repositórios...');
    if (username) {
      setRepos(await GithubService.getRepos(username));
    }
    setLoading(false);
  };

  useEffect(() => {
    getRepos();

    return () => {
      setRepos([]);
    };
  }, [username]);

  return (
    <>
      <div className='row'>
        <h2>
          Lista de Repositórios de{' '}
          <a href='#' onClick={handleNavigateUser}>
            {username}
          </a>
        </h2>
      </div>
      <div className='row'>
        <div className='col-12'>
          <RepoList repos={repos} />
        </div>
      </div>
    </>
  );
};

export default ReposPage;
