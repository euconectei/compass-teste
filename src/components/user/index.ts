import UserDisplay from './UserDisplay';
import UserNotFound from './UserNotFound';

export { UserDisplay, UserNotFound };
