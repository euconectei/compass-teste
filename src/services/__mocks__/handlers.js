import { rest } from 'msw';
import { reposMock } from '../../types/__mocks__/repos.mock';
import { starredsMock } from '../../types/__mocks__/starreds.mock';
import { userMock } from '../../types/__mocks__/user.mock';

export const handlers = [
  rest.get('https://api.github.com/users/:user/repos', (req, res, ctx) => {
    return res(ctx.json(reposMock));
  }),
  rest.get('https://api.github.com/users/:user/starred', (req, res, ctx) => {
    return res(ctx.json(starredsMock));
  }),
  rest.get('https://api.github.com/users/:user', (req, res, ctx) => {
    return res(ctx.json(userMock));
  }),
];
