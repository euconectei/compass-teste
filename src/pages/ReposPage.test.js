import { fireEvent, render, screen } from '@testing-library/react';
import { rest } from 'msw';

import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';
import renderer from 'react-test-renderer';
import AuthProvider from '../contexts/AuthContext';
import LoadingProvider from '../contexts/LoadingContext';
import GithubService from '../services/GithubService';
import { server } from '../services/__mocks__/server';
import Layout from '../templates/Layout';
import { reposMock } from '../types/__mocks__/repos.mock';
import { userMock } from '../types/__mocks__/user.mock';

import ReposPage from './ReposPage';

const userExisting = 'euconectei';
const userNotExisting = 'euconectei123';

const Providers = ({ children }) => {
  return (
    <>
      <LoadingProvider>
        <AuthProvider>{children}</AuthProvider>
      </LoadingProvider>
    </>
  );
};

const renderWithProviders = (ui) => render(ui, { wrapper: Providers });

describe('Testing ReposPage Page', () => {
  it('should render the ReposPage component', () => {
    const page = renderer.create(
      <BrowserRouter>
        <ReposPage />
      </BrowserRouter>
    );
    expect(page).toMatchSnapshot();
  });

  it.only('should render the ReposPage component with a user', async () => {
    // jest.useFakeTimers();

    server.use(
      rest.get(
        `https://api.github.com/users/${userExisting}/repos`,
        (req, res, ctx) => {
          return res(ctx.delay(300), ctx.json(reposMock));
        }
      )
    );

    const { debug, rerender } = renderWithProviders(
      <MemoryRouter initialEntries={[`/${userExisting}/repos`]}>
        <Routes>
          <Route path='/' element={<Layout />}>
            <Route path='/:username/repos' element={<ReposPage />} />
          </Route>
        </Routes>
      </MemoryRouter>
    );

    // debug();

    // localhost:3000/:username/repos

    // testar se o loading esta funcionando
    // jest.advanceTimersByTime(10);
    expect(screen.getByText('Carregando')).toBeInTheDocument();
    // testar se o usuario está sendo renderizado
    expect(screen.getByText(userExisting)).toBeInTheDocument();
    // jest.advanceTimersByTime(3000);
    // rerender();
    debug();
    screen.waitFor(() => {
      expect(screen.getByText(reposMock[0].name)).toBeInTheDocument();
    });
  });

  it('should render the ReposPage component with a user not found', async () => {
    GithubService.getUser = jest.fn().mockResolvedValue(userNotFound);

    render(
      <BrowserRouter>
        <ReposPage />
      </BrowserRouter>
    );

    const userInput = screen.getByPlaceholderText('Usuário do Github');
    fireEvent.change(userInput, { target: { value: userNotFound } });

    const submitButton = screen.getByText('Buscar');
    fireEvent.click(submitButton);

    expect(userInput.value).toBe(userExisting);
    await wait(() =>
      expect(screen.getByText('Usuário não encontrado')).toBeInTheDocument()
    );
  });
});
