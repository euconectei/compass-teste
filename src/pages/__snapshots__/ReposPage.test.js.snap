// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ReposPage Page should render the ReposPage component 1`] = `
Array [
  <div
    className="row"
  >
    <h2>
      Lista de Repositórios de
       
      <a
        href="#"
        onClick={[Function]}
      />
    </h2>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-12"
    >
      <div
        className="alert alert-info"
      >
        Nenhum repositório encontrado.
      </div>
    </div>
  </div>,
]
`;
